d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java

```diff
@@ -0,0 +1,391 @@
+   * @param oldPartitionState The old state of the partition
+   * @param clusterAvgLatency The average latency of the cluster of last interval
+   */
+  private void enrollNewQuarantineAndRecovery(
+      PartitionState newPartitionState,
+      PartitionState oldPartitionState, long clusterAvgLatency)
+  {
+    int partitionId = newPartitionState.getPartitionId();
+    Map<TrackerClient, LoadBalancerQuarantine> quarantineMap = newPartitionState.getQuarantineMap();
+    Map<TrackerClient, LoadBalancerQuarantine> quarantineHistory = newPartitionState.getQuarantineHistory();
+    Set<TrackerClient> recoverySet = newPartitionState.getRecoveryTrackerClients();
+
+    for (TrackerClient trackerClient : newPartitionState.getTrackerClients())
+    {
+      TrackerClientState trackerClientState = newPartitionState.getTrackerClientStateMap().get(trackerClient);
+
+      double serverWeight = trackerClient.getPartitionWeight(partitionId);
+      // Check and enroll quarantine map
+      boolean isQuarantined = enrollClientInQuarantineMap(trackerClient, trackerClientState, serverWeight, quarantineMap,
+          quarantineHistory, newPartitionState.getTrackerClientStateMap().size(), clusterAvgLatency);
+      // check and enroll recovery set
+      enrollClientInRecoverySet(isQuarantined, trackerClient, trackerClientState, serverWeight, recoverySet,
+          oldPartitionState);
+    }
+  }
+
+  /**
+   * Perform fast recovery or regular recovery when necessary
+   * Fast recovery will double the current health score when it is enabled
+   *
+   * @param trackerClient The {@link TrackerClient} to be recovered
+   * @param trackerClientState The state of the {@link TrackerClient}
+   * @param recoverySet A set of {@link TrackerClient} to be recovered
+   */
+  private void handleClientInRecoverySet(TrackerClient trackerClient, TrackerClientState trackerClientState,
+      Set<TrackerClient> recoverySet)
+  {
+    if (trackerClientState.getCallCount() < trackerClientState.getAdjustedMinCallCount())
+    {
+      double healthScore = trackerClientState.getHealthScore();
+      if (healthScore <= RelativeStateUpdater.MIN_HEALTH_SCORE)
+      {
+        // Reset the health score to initial recovery health score if health score dropped to 0 before
+        trackerClientState.setHealthScore(INITIAL_RECOVERY_HEALTH_SCORE);
+      }
+      else if (_fastRecoveryEnabled)
+      {
+        // If fast recovery is enabled, we perform fast recovery: double the health score
+        healthScore *= FAST_RECOVERY_FACTOR;
+        trackerClientState.setHealthScore(Math.min(healthScore, RelativeStateUpdater.MAX_HEALTH_SCORE));
+      }
+    }
+    else if (!_fastRecoveryEnabled
+        || !trackerClientState.isUnhealthy()
+        || trackerClientState.getHealthScore() > FAST_RECOVERY_HEALTH_SCORE_THRESHOLD)
+    {
+      /**
+       * Remove the client from the map if:
+       * 1. fast recovery is not enabled OR
```

Comment:
I just changed the 2nd logic, I think the previous 2nd condition was actually wrong.
else if (!_fastRecoveryEnabled
        || trackerClientState.isUnhealthy()
        || trackerClientState.getHealthScore() > FAST_RECOVERY_HEALTH_SCORE_THRESHOLD)
    {
      /**
       * Remove the client from the map if:
       * 1. fast recovery is not enabled OR
       * 2. the client is still unhealthy OR
       * 3. The health score is beyond 0.5, we will let it perform normal recovery
       */
      recoverySet.remove(trackerClient);
    }

I think it is necessary to check the fast recovery, only if fast recovery is enabled, it makes sense to keep it in the recovery set.
For example, one host is quarantined, if fast recovery is not enabled, its health score can be recovered in the pattern: 0.00, 0.01, 0.06, 0.11....
if fast recovery is enabled, the health score recovery pattern can be 0.00, 0.01, 0.02, 0.04, 0.08, 0.16 if it's not getting enough traffic.

There is no point to keep a host in the recovery set if fast recovery is not enabled. Because even if it's not getting enough traffic, we cannot perform fast recovery for the host.

Reply:
Will remove the fast recovery check here, instead, we will add the fast recovery check when we add a new client to the recovery set.